Wrap many element components in lib with memo

lukasmasuch opened the pull request:
## Describe your changes

This PR adds many of our element components with `memo` to prevent unnecessary rerenders. For some components, it also changes it to use the `useTheme` hook instead of `withTheme`.

## Testing Plan

- No logical or visual changes. Existing tests should be sufficient.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch committed:
Add memo to number input

lukasmasuch committed:
Add memo to a lot more elements

lukasmasuch committed:
Fix form import

lukasmasuch committed:
Remove export

lukasmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

sfc-gh-bnisco commented on frontend/lib/src/components/elements/Toast/Toast.tsx:
```diff
@@ -15,14 +15,15 @@
   useState,
 } from "react"
 
-import { withTheme } from "@emotion/react"
```
**suggestion (non-blocking):** Since we're leaning into hooks, it would be great to add an eslint rule that prevents usage of `withTheme` and suggests using `useTheme` instead. Totally fine if this is a follow-up, since it's not exactly the point of this PR.

lukasmasuch replied:
Sounds good 👍 I added a rule to prevent usage of `withTheme`.

sfc-gh-bnisco reviewed (approved):
Code LGTM. Generally speaking, it would be great if we had a way to measure the impact of this change (that's part of the reason why the perf project is important!).

Medium term: it would be good to come up with guidance/automation on these kinds of paradigms within Streamlit so that we have a structure that everyone can follow. Perhaps some kind of eslint rule that ensures any exported React component from the `lib/src/components` directory utilizes `memo`.

Longer term: React compiler is probably going to make most of these changes unnecessary

lukasmasuch committed:
Add eslint rule to prevent usage of withTheme

lukasmasuch committed:
Remove unnecessary import

lukasmasuch commented:
> it would be great if we had a way to measure the impact of this change

@sfc-gh-bnisco Good point, happy to wait with merging this until we have some more metrics. Would https://github.com/streamlit/streamlit/pull/9769 already give us some metrics for this?

Also, I think just adding memo everywhere will reduce some rerenders.. but I'm not expecting too much of an impact for reruns. The element proto objects will change the reference with every rerun even if its the same proto content -> which requires a new render of every element. Added a tech debt item for this [here](https://www.notion.so/snowflake-corp/Use-custom-memo-dependency-that-checks-on-element-proto-hash-13f7170bb4168081beebddfae1fc714f). 

> Perhaps some kind of eslint rule that ensures any exported React component from the lib/src/components directory utilizes memo

+1. I added a tech debt item [here](https://www.notion.so/snowflake-corp/Add-eslint-rule-to-enforce-usage-of-memo-for-all-lib-components-13f7170bb416804c8bb0dc99172d9c01).

sfc-gh-bnisco commented:
@lukasmasuch I do not expect https://github.com/streamlit/streamlit/pull/9769 to give us a meaningful information on this change, since this change is more impactful in re-renders, and lighthouse is mostly around initial load times. This would be best captured by later work items in the Perf workstream, specifically Playwright performance tests.

Thanks for filing that tech debt item about element proto objects. I think that is a great candidate for something we should have a clear ability to measure before tackling. My gut instinct is that that will also be measured via Playwright performance.

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

lukasmasuch committed:
Add memo import

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

lukasmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

sfc-gh-lmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

lukasmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

sfc-gh-lmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

lukasmasuch commented:
@sfc-gh-bnisco, is it a good time to merge this PR? If the `memo` uses impact overall performance, we will likely be able to measure it to some degree in our new performance metrics.

lukasmasuch committed:
Merge branch 'develop' into add-memo-for-number-input

sfc-gh-bnisco commented:
@lukasmasuch You're certainly unblocked to do so! I'd expect it to reflect in the existing Playwright tests. Another option is to first add even more test cases to get a baseline for areas you think will be particularly impacted, and then check this in. Up to you though!

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into add-memo-for-number-input

lukasmasuch committed:
Remove withTheme from sidebar

lukasmasuch committed:
More fixes for sidebar

lukasmasuch commented:
ok, I think I will merge as is. The added memo's might prevent maybe a couple of unnecessary rerenders, but I don't expect anything significant regarding performance. And one of the other reasons for this PR is to add a bit more consistency to how all our components behave and a bit less about performance.
